src/pudl/helpers.py

```diff
@@ -1476,6 +1476,28 @@ def dedupe_and_drop_nas(
     )
 
 
+def drop_records_with_null_pk(
+    df: pd.DataFrame, primary_key_col: str, num_of_expected_nulls: int
+) -> pd.DataFrame:
+    """Drop a prescribed number of records with null values in a primary key column.
+
+    Args:
+        df: table with primary_key_col column.
+        primary_key_col: name of column which potential null values.
+        num_of_expected_nulls: the number of records
+
+    Raises:
+        AssertionError: If there are more nulls in the df then the
+            num_of_expected_nulls.
+    """
+    # there is one record that has a null gen id. ensure there isn't more before dropping
+    if len(null_gens := df[df[primary_key_col].isnull()]) > num_of_expected_nulls:
+        raise AssertionError(
+            f"Expected {num_of_expected_nulls} or zero records with a null {primary_key_col} but found {null_gens}"
```

Comment:
What does the "or zero records with a null" in here mean? Is there some reason why this wouldn't be fine with an expectation of zero null values?

Reply:
if there are zero records will nulls in it then this function will effectively do nothing so i don't think it would be a problem. if there are zero null records than perhaps we could stop even trying to drop things so raising a warning could be helpful but we'd need to manage the num_of_expected_nulls differently for fast or full etl runs and that does not seem worth the candle.

Reply:
but i think the proper thing to say here is that we "Expected `num_of_expected_nulls` or less records". which... if `num_of_expected_nulls` == 1 than that will imply zero

Reply:
i've updated the message and comment